shell/browser/api/electron_api_utility_process.cc

```diff
@@ -210,37 +210,58 @@ UtilityProcessWrapper::UtilityProcessWrapper(
   network_context->CreateHostResolver(
       {}, host_resolver.InitWithNewPipeAndPassReceiver());
   params->host_resolver = std::move(host_resolver);
+
   node_service_remote_->Initialize(std::move(params));
+  node_service_remote_->SetTermination(base::BindOnce(
+      &UtilityProcessWrapper::HandleTermination, weak_factory_.GetWeakPtr()));
 }
 
-UtilityProcessWrapper::~UtilityProcessWrapper() = default;
+UtilityProcessWrapper::~UtilityProcessWrapper() {
+  content::ServiceProcessHost::RemoveObserver(this);
+}
 
-void UtilityProcessWrapper::OnServiceProcessLaunched(
+void UtilityProcessWrapper::OnServiceProcessLaunch(
     const base::Process& process) {
   DCHECK(node_service_remote_.is_connected());
   pid_ = process.Pid();
   GetAllUtilityProcessWrappers().AddWithID(this, pid_);
-  if (stdout_read_fd_ != -1) {
+  if (stdout_read_fd_ != -1)
     EmitWithoutEvent("stdout", stdout_read_fd_);
-  }
-  if (stderr_read_fd_ != -1) {
+  if (stderr_read_fd_ != -1)
     EmitWithoutEvent("stderr", stderr_read_fd_);
-  }
-  // Emit 'spawn' event
   EmitWithoutEvent("spawn");
 }
 
-void UtilityProcessWrapper::OnServiceProcessDisconnected(
-    uint32_t error_code,
-    const std::string& description) {
+void UtilityProcessWrapper::HandleTermination(uint64_t exit_code) {
   if (pid_ != base::kNullProcessId)
     GetAllUtilityProcessWrappers().Remove(pid_);
   CloseConnectorPort();
-  // Emit 'exit' event
-  EmitWithoutEvent("exit", error_code);
+
+  v8::Isolate* isolate = JavascriptEnvironment::GetIsolate();
+  if (node::Environment::GetCurrent(isolate))
```

Comment:
What is the reason for this check ?

Reply:
@deepak1556 it can be called after the environment has been torn down, leading to the following crash:

```
  #  /workspaces/gclient/src/out/Testing/electron[14295]: MaybeLocal<Value> node::MakeCallback(Isolate *, Local<Object>, Local<String>, int, Local<Value> *, async_context) at ../../third_party/electron_node/src/api/callback.cc:252
  #  Assertion failed: (env) != nullptr

----- Native stack trace -----

 1: 0x5fb877bd0a2e  [/workspaces/gclient/src/out/Testing/electron electron/spec]
 2: 0x5fb877df7277  [/workspaces/gclient/src/out/Testing/electron electron/spec]
 3: 0x5fb877ba36bd node::MakeCallback(v8::Isolate*, v8::Local<v8::Object>, v8::Local<v8::String>, int, v8::Local<v8::Value>*, node::async_context) [/workspaces/gclient/src/out/Testing/electron electron/spec]
 4: 0x5fb84eb8e0ce  [/workspaces/gclient/src/out/Testing/electron electron/spec]
 5: 0x5fb84e721272  [/workspaces/gclient/src/out/Testing/electron electron/spec]
 6: 0x5fb84e712eda  [/workspaces/gclient/src/out/Testing/electron electron/spec]
 7: 0x5fb85e8cd0be  [/workspaces/gclient/src/out/Testing/electron electron/spec]
 8: 0x5fb85ee4f98c  [/workspaces/gclient/src/out/Testing/electron electron/spec]
 9: 0x5fb85ee4ffc4  [/workspaces/gclient/src/out/Testing/electron electron/spec]
10: 0x5fb85cfd5ccd  [/workspaces/gclient/src/out/Testing/electron electron/spec]
11: 0x5fb85d0f1ffb  [/workspaces/gclient/src/out/Testing/electron electron/spec]
12: 0x5fb85d0e5279  [/workspaces/gclient/src/out/Testing/electron electron/spec]
13: 0x5fb85d0edbce  [/workspaces/gclient/src/out/Testing/electron electron/spec]
14: 0x5fb85d0da8fb  [/workspaces/gclient/src/out/Testing/electron electron/spec]
15: 0x5fb84f2f8477  [/workspaces/gclient/src/out/Testing/electron electron/spec]
16: 0x5fb84f2fd147  [/workspaces/gclient/src/out/Testing/electron electron/spec]
17: 0x5fb84f2fc636  [/workspaces/gclient/src/out/Testing/electron electron/spec]
18: 0x5fb84f2f5024  [/workspaces/gclient/src/out/Testing/electron electron/spec]
19: 0x5fb84f2f5923  [/workspaces/gclient/src/out/Testing/electron electron/spec]
20: 0x5fb84e4b9223  [/workspaces/gclient/src/out/Testing/electron electron/spec]
21: 0x78146c29c083 __libc_start_main [/lib/x86_64-linux-gnu/libc.so.6]
22: 0x5fb84e39102a _start [/workspaces/gclient/src/out/Testing/electron electron/spec]
```

Reply:
Can we move the check into the event emitter api https://github.com/electron/electron/blob/b74d3e14b004d751f1920efae62f730c98bce344/shell/common/gin_helper/event_emitter_caller.cc#L12 ?

Reply:
sure, i'll try that

Reply:
Done in 52dc3d442d